apps/common/services.py

```diff
@@ -0,0 +1,111 @@
+
+from apps.common.db import Base
+
+ModelType = TypeVar("ModelType", bound=Base)
+CreateSchemaType = TypeVar("CreateSchemaType", bound=BaseModel)
+ReadSchemaType = TypeVar("ReadSchemaType", bound=BaseModel)
+UpdateSchemaType = TypeVar("UpdateSchemaType", bound=BaseModel)
+DeleteSchemaType = TypeVar("DeleteSchemaType", bound=BaseModel)
+
+
+class AsyncCRUDBase:
+    def __init__(self, *, model: Type[ModelType]):
+        self.model = model
+
+    async def create(self, *, session: AsyncSession, data: Union[Dict, None] = None,
+                     obj: Union[CreateSchemaType, None] = None) -> ModelType:
+        data = data if data else {}
+        obj_in_data = jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False) if obj else {}
+        insert_statement = insert(self.model).values(**data, **obj_in_data).returning(self.model)
+        statement = select(self.model).from_statement(insert_statement).execution_options(populate_existing=True)
+        result: ChunkedIteratorResult = await session.execute(statement=statement)
+        await session.commit()
+        data: ModelType = result.scalar_one()
+        return data
+
+    async def create_many(self, *, session: AsyncSession, objs: List[CreateSchemaType]) -> Union[List[ModelType], None]:
+        insert_statement = (
+            insert(self.model)
+                .values([jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False) for obj in objs])
+                .returning(self.model)
+        )
+        statement = select(self.model).from_statement(insert_statement).execution_options(populate_existing=True)
+        result: ChunkedIteratorResult = await session.execute(statement=statement)
+        await session.commit()
+        data: Union[List[ModelType], None] = result.scalars().all()
+        return data
+
+    async def read(self, *, session: AsyncSession, data: Union[Dict, None] = None,
+                   obj: Union[ReadSchemaType, None] = None) -> Union[ModelType, None]:
+        data = data if data else {}
+        obj_in_data = jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False) if obj else {}
+        where_dict = {**data, **obj_in_data}
+        statement = select(self.model)
+        model = self.model.c if isinstance(self.model, Table) else self.model
+        for k, v in where_dict.items():
+            statement = statement.where(getattr(model, k) == v)
+        result: ChunkedIteratorResult = await session.execute(statement=statement)
+        response: Union[ModelType, None] = result.first()
+        return response if isinstance(self.model, Table) else response[0]
+
+    async def update(self, *, session: AsyncSession, data: Union[Dict, None] = None,
+                     obj: Union[Dict, UpdateSchemaType] = None) -> Union[ModelType, None]:
+        obj = obj if obj else {}
+        data = data if data else {}
+        values = obj if isinstance(obj, Dict) else jsonable_encoder(obj=obj, exclude_unset=True, by_alias=False)
```

Comment:
Check jsonable_encoder, though it works with the dict type automatically.

Reply:
Ok